### PR TITLE
util/winutil/winenv: add package for current Windows environment details

### PR DESCRIPTION
**Pull request by nickkhyl**
Package winenv provides information about the current Windows environment. This includes details such as whether the device is a server or workstation, and if it is AD domain-joined, MDM-registered, or neither.

Updates tailscale/corp#18342

### CODE REVIEW

File: util/winutil/winenv/mksyscall.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+package winenv
+
+//go:generate go run golang.org/x/sys/windows/mkwinsyscall -output zsyscall_windows.go mksyscall.go
+
+// https://web.archive.org/web/20240407040123/https://learn.microsoft.com/en-us/windows/win32/api/mdmregistration/nf-mdmregistration-isdeviceregisteredwithmanagement
+//sys isDeviceRegisteredWithManagement(isMDMRegistered *bool, upnBufLen uint32, upnBuf *uint16) (hr int32, err error) = MDMRegistration.IsDeviceRegisteredWithManagement?
+
+// https://web.archive.org/web/20240407035921/https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-verifyversioninfow
+//sys verifyVersionInfo(verInfo *osVersionInfoEx, typ verTypeMask, cond verCondMask) (res bool) = VerifyVersionInfoW
```
Comment (dblohm7) [inline util/winutil/winenv/mksyscall.go:11]: Please be explicit with the module name.

File: util/winutil/winenv/mksyscall.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+// https://web.archive.org/web/20240407035706/https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-versetconditionmask
+//sys verSetConditionMask(condMask verCondMask, typ verTypeMask, cond verCond) (res verCondMask) = VerSetConditionMask
```
Comment (dblohm7) [inline util/winutil/winenv/mksyscall.go:14]: And here

File: util/winutil/winenv/winenv_windows.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+// osVersionInfoEx contains operating system version information.
+// See [OSVERSIONINFOEXW] for details.
+//
+// [OSVERSIONINFOEXW]: https://web.archive.org/web/20240407035213/https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-osversioninfoexw
+type osVersionInfoEx struct {
```
Comment (dblohm7) [inline util/winutil/winenv/winenv_windows.go:20]: Just reuse the definition in `x/sys/windows`?

File: util/winutil/winenv/winenv_windows.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	_VER_GREATER       = verCond(2)
+	_VER_GREATER_EQUAL = verCond(3)
+	_VER_LESS          = verCond(4)
+	_VER_LESS_EQUAL    = verCond(5)
+	_VER_AND           = verCond(6)
+	_VER_OR            = verCond(7)
+)
+
+// IsDomainJoined reports whether the device is domain-joined.
+func IsDomainJoined() bool {
```
Comment (dblohm7) [inline util/winutil/winenv/winenv_windows.go:69]: Oh, nice! I think I might need to use this in some other stuff I've been playing with. Could you please move this function into `util/winutil/winutil_windows.go`?

**Review by dblohm7 — APPROVED** [review]
OK with comments addressed.

File: util/winutil/winenv/winenv_windows.go (hunk 4 above)
Comment (nickkhyl) [reply util/winutil/winenv/winenv_windows.go:69]: How we we feel about moving all this stuff into `util/winutil`? Originally, I intended `util/winutil/winenv` to be a small reusable package focused on providing information about the current Windows environment. But if we'd like to have `IsDomainJoined` moved to `util/winutil`, we might as well move all this stuff there. I really see no reason to have `IsDomainJoined`, `IsMDMRegistered` and similar functions in different packages.

Alternatively, we could move the `winenv` package into `util/`, as there's really no dependency between `winenv` and `winutil`.

Thoughts?

Comment (dblohm7) [reply util/winutil/winenv/winenv_windows.go:69]: Actually, I changed my mind on this one. @nickkhyl don't worry about it! :-)

File: util/winutil/winenv/mksyscall.go (hunk 1 above)
Comment (nickkhyl) [reply util/winutil/winenv/mksyscall.go]: Done.

File: util/winutil/winenv/mksyscall.go (hunk 2 above)
Comment (nickkhyl) [reply util/winutil/winenv/mksyscall.go]: Done.

**Pull request merged**